picamera2/controls.py

```diff
@@ -0,0 +1,160 @@
+import libcamera
+
+
+class Controls():
+    """ A class for manipulating picamera2 controls.
+
+    All instances of this class within the picamera2 module share the same
+    values. Before the camera is started, whatever the user sets will
+    take effect when the camera starts. Based on how the CompletedRequests
+    class is structured, any reassignment of a control will take effect on
+    the next request.
+
+    Controls are type checked just before being passed to the
+    camera manager. The user must supply the exact syntax for a control and
+    its value as it is found in libcamera when creating the control.
+    """
+
+    AeConstraintMode: (int, str) = None
+    AeEnable: (int, bool) = None
+    AeExposureMode: (int, str) = None
+    AeMeteringMode: (int, str) = None
+    AePrecaptureTrigger: int = None
+    AfTrigger: int = None
+    AnalogueGain: (float, int) = None
+    AwbEnable: (int, bool) = None
+    AwbMode: (int, str) = None
+    Brightness: (float, int) = None
+    ColorCorrectionAberrationMode: (int, str) = None
+    ColourCorrectionMatrix: (tuple, list) = None
+    ColourGains: (tuple, list) = None
+    Contrast: (float, int) = None
+    DigitalGain: (float, int) = None
+    ExposureTime: (float, int) = None
+    ExposureValue: (float, int) = None
+    FrameDurationLimits: (tuple, list) = None
+    MaxLatency: int = None
+    NoiseReductionMode: (int, str) = None
+    PipelineDepth: int = None
+    Saturation: (float, int) = None
+    ScalerCrop: (tuple, list) = None
+    SceneFlicker: (int, str) = None
+    Sharpness: (float, int) = None
+    TestPatternMode: (int, str) = None
+
```

Comment:
One thing that's bothering me a bit is that we have a hard-coded list of stuff here, and how we are going to maintain it. I'm thinking of libcamera adding new controls, or changing the definitions of existing controls. Will they work immediately (as they do now) or will we have to update Picamera2 to "catch up"? And who does the "catching up" and how quickly will they do it? Could there be a way of supporting "unfamiliar" controls?

Reply:
I'm also bothered by this. If there is a yaml file on the libcamera end that can be parsed, then that would be a solution. In theory, they would continue to work as long as we know what the name is and what types it can be.

Reply:
Just to add to this, what would the effort be like to add all of the camera controls to the Picamera2().camera.controls output? Presumably that comes from a place libcamera, so that would also be source for the current camera settings and would reduce maintenance effort in the future..